kubernetes/samples/scripts/domain-lifecycle/helper.sh

```diff
@@ -0,0 +1,238 @@
+
+  replicasCmd="(.spec.clusters[] \
+    | select (.clusterName == \"${clusterName}\")).replicas"
+  maxReplicaCmd="(.status.clusters[] | select (.clusterName == \"${clusterName}\")) \
+    | .maximumReplicas"
+  replica=$(echo ${domainJson} | jq "${replicasCmd}")
+  if [[ -z "${replica}" || "${replica}" == "null" ]]; then
+    replica=$(echo ${domainJson} | jq .spec.replicas)
+  fi
+  if [ "${operation}" == "DECREMENT" ]; then
+    replica=$((replica-1))
+    if [ ${replica} -lt 0 ]; then
+      replica=0
+    fi
+  elif [ "${operation}" == "INCREMENT" ]; then
+    replica=$((replica+1))
+    maxReplicas=$(echo ${domainJson} | jq "${maxReplicaCmd}")
+    if [ ${replica} -gt ${maxReplicas} ]; then
+      echo "${errorMessage}"
+      eval $__result="MAX_REPLICA_COUNT_EXCEEDED"
+      return
+    fi
+  fi
+
+  cmd="(.spec.clusters[] | select (.clusterName == \"${clusterName}\") \
+    | .replicas) |= ${replica}"
+  replicaPatch=$(echo ${domainJson} | jq "${cmd}" | jq -cr '(.spec.clusters)')
+  eval $__result="'${replicaPatch}'"
+  eval $__replicaCount="'${replica}'"
+}
+
+#
+# Function to validate whether a server belongs to a  cluster or is an independent managed server
+# $1 - Domain unique id.
+# $2 - Domain namespace.
+# $3 - Return value indicating if server is valid (i.e. if it's part of a cluster or independent server).
+# $4 - Retrun value containting cluster name to which this server belongs.
+#
+function validateServerAndFindCluster {
+  local domainUid=$1
+  local domainNamespace=$2 
+  local __isValidServer=$3
+  local __clusterName=$4
+  local errorMessage="Server name is outside the range of allowed servers. \
+Please make sure server name is correct."
+
+  configMap=$(${kubernetesCli} get cm ${domainUid}-weblogic-domain-introspect-cm \
+    -n ${domainNamespace} -o json)
+  topology=$(echo "${configMap}" | jq '.data["topology.yaml"]')
+  jsonTopology=$(python -c \
+    'import sys, yaml, json; print json.dumps(yaml.safe_load('"${topology}"'), indent=4)')
+  servers=($(echo $jsonTopology | jq -r '.domain.servers[].name'))
+  if  checkStringInArray "${serverName}" "${servers[@]}" ; then
+    eval $__clusterName=""
+    eval $__isValidServer=true
+  else
+    dynamicClause=".domain.configuredClusters[] | select (.dynamicServersConfig != null)"
+    namePrefixSize=". | {name: .name, prefix:.dynamicServersConfig.serverNamePrefix, \
+                 max:.dynamicServersConfig.maxDynamicClusterSize}"
+    dynamicClusters=($(echo $jsonTopology | jq "${dynamicClause}" | jq -cr "${namePrefixSize}"))
+    for dynaClusterNamePrefix in "${dynamicClusters[@]}"; do
+      prefix=$(echo ${dynaClusterNamePrefix} | jq -r .prefix)
+      if [[ "${serverName}" == "${prefix}"* ]]; then
+        serverCount=$(echo "${serverName: -1}")
+        maxSize=$(echo ${dynaClusterNamePrefix} | jq -r .max)
+        if [ ${serverCount} -gt ${maxSize} ]; then
+          printError "${errorMessage}"
```

Comment:
Out of curiosity, can this ever happen? Or is it more in the nature of an assertion?

Reply:
Yes. For e.g. let's say "max dynamic cluster size" of a dynamic cluster is 4 and server-name prefix is "managed-server". In this case, only managed-server1 through managed-server4 can be started. If you try to start "managed-server5" or "managed-server6", they won't start.